File: weeks/8/2/solution.py
import sys
sys.setrecursionlimit(100000)
readline = sys.stdin.readline

numPeaks = int(readline())
peaks = map(int, readline().split())

peakStack = []

for peak in peaks:
    print(len(peakStack), end=" ")

    # 현재 봉우리보다 작은 봉우리를 스택에서 버림
    while True:
        if len(peakStack) == 0:
            break
        if peakStack[-1] > peak:
            break

        peakStack.pop()

    peakStack.append(peak)

print("")

File: weeks/8/2/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef long long ll;

void solve() {
  int numPeaks;
  cin >> numPeaks;

  vector<int> peakStack;

  for (int i = 0; i < numPeaks; ++i) {
    int peak;
    cin >> peak;

    // 현재 스택의 개수 출력
    cout << (int)(peakStack.size()) << " ";

    // 현재 봉우리보다 작은 봉우리를 스택에서 버림
    while (true) {
      if (peakStack.empty()) {
        break;
      }
      if (peak < peakStack.back()) {
        break;
      }

      peakStack.pop_back();
    }

    // 현재 봉우리를 스택에 넣음
    peakStack.push_back(peak);
  }

  cout << '\n';
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  solve();
}
